core/trino-main/src/main/java/io/trino/operator/scalar/JoniRegexpFunctions.java

```diff
@@ -408,4 +409,14 @@ public static long regexpCount(@SqlType("varchar(x)") Slice source, @SqlType(Jon
 
         return count;
     }
+
+    public static int getSearchingOffset(Matcher matcher, int at, int range)
+    {
+        try {
+            return matcher.searchInterruptible(at, range, Option.DEFAULT);
+        }
+        catch (InterruptedException interruptedException) {
+            throw new TrinoException(GENERIC_USER_ERROR, "Regexp matching interrupted");
```

Comment:
Recover interrupted flag

Reply:
This should not be "user error". In fact, it should probably not be a TrinoException at all. If this triggers, the query should've already been canceled or aborted and this will be invisible to users.

Reply:
@martint what mechanism would trigger the cancel/abort? theoretically, if `searchInterruptible` running for say 600s,  wouldn't the query still be alive until we kill using https://github.com/trinodb/trino/pull/12392/  (unless some other issues with the query)?

Reply:
If I remember correctly, at some point there used to be code to interrupt threads when a query is canceled or fails (or maybe we just talked about it). Not sure if that is still there, but if not, then this change won't do anything until such mechanism is reinstated.

Reply:
Hey, this is the statechange listener's code when remove the task:

```
 private void doRemoveTask(TaskHandle taskHandle)
    {
        List<PrioritizedSplitRunner> splits;
        synchronized (this) {
            tasks.remove(taskHandle);
            splits = taskHandle.destroy();

            // stop tracking splits (especially blocked splits which may never unblock)
            allSplits.removeAll(splits);
            intermediateSplits.removeAll(splits);
            blockedSplits.keySet().removeAll(splits);
            waitingSplits.removeAll(splits);
        }

        // call destroy outside of synchronized block as it is expensive and doesn't need a lock on the task executor
        for (PrioritizedSplitRunner split : splits) {
            split.destroy();
        }

        // record completed stats
        long threadUsageNanos = taskHandle.getScheduledNanos();
        completedTasksPerLevel.incrementAndGet(computeLevel(threadUsageNanos));

        log.debug("Task finished or failed %s", taskHandle.getTaskId());
    }
```

We found it didn't modify the running split, so https://github.com/trinodb/trino/pull/12392 added a periodic job to evaluate each running split to make sure they don't stuck. 

Do you have suggestion as what exception to throw? Maybe GENERIC_INTERNAL_ERROR or EXCEEDED_TIME_LIMIT?

Reply:
I am thinking lower the timeout threholds to 10s from 10minutes, so user can be notified for this issue. So they can stop sending those complex regular expressions. We should also prevent task level retry from retrying this function.

Reply:
Recover interrupted flag:
This is already handled by the JONI library after checking the interrupt flag.

Reply:
> Recover interrupted flag: This is already handled by the JONI library after checking the interrupt flag.

Please check my response above

Reply:
From what I understand we want to preserve the classification of the given error as a `USER_ERROR`, as :

- For the cases when a Driver is interrupted due to a task failure or query cancellation the type of error doesn't matter
- https://github.com/trinodb/trino/pull/12392 introduces background process that will interrupt drivers that run for more than 10 minutes. In such case the exception thrown will be propagated to the user and must be treated as `USER_ERROR`.

@leetcode-1533 Is my understanding correct?

Reply:
For the second case, the query should be aborted as a first class action, not as a side effect of tasks being interrupted, so the error type and reason can be indicated at that point.

Reply:
makes sense. So the way I understand is that, as part of periodic interruption in #12392, we should check if the query context still exists, and fail the query with a proper exception there. 

We also know the stacktrace for the thread we're interrupting there, so should be able to provide _some_ information around what exactly ran long.